ls1intum/Artemis
# [Bug] SCA categorization when build fails

### Checklist

### Motivation and Context


PMD and Checkstyle are in some cases able to deliver feedback even if compilation fails (e.g. on a type error). In such a case the feedback was not properly categorized.

### Description


- Move categorization functionality to `StaticCodeAnalysisService`
- Call categorization functionality even if no test case feedback exists
- Fixed typos in some category mappings
- Move Spotbugs `I18N` ruleset (usage of locale and encodings) to category best practices
- Fix the result String generation for Jenkins. Use the test case feedback to determine the String not all feedback.
- Tests

### Steps for Testing


**Test the Bamboo and Jenkins setup**

1. Log in to Artemis
2. Create a programming exercise with static code analysis enabled
3. Participate in the exercise
4. Submit and check that feedback is properly categorized (in the default configuration, documentation issues should not show up)
5. Let the build fail by submitting the following implementation for a sort algorithm and check that the static code analysis feedback is properly categorized.
```
Collections.sort(input);
int a = "";
```

### Test Coverage

Coverage is at the same level as a method was just moved to another service and called at a slightly different spot. Added two more specialized tests which don't improve coverage.
